Create reading-week-checklist.md (#1003)
foundersandcoders/master-reference: coursebook/general/reading-week-checklist.md

- [ ] When we make a change to a file, how do we tell git to track it?
- [ ] When we have finished working on a branch, how do we make sure that our changes do not cause a conflict with master? (this can all be done **locally**)
- [ ] What does `git push origin [branch-name]` do?
- [ ] Why do we make pull requests instead of just changing master directly?
- [ ] Why is it important to run our team member's branches when they make a pull request?
- [ ] This is the working process you should follow, try to explain what happens at each step/why we do it:
    - `git clone [repository]`
    - `git checkout -b new-feature`
    *make some changes to index.js*
    - `git add index.js`
    - `git commit -m "added a cool new feature"`
    *since we started, our partner has merged another branch into master*
    - `git checkout master`
    - `git pull origin master`
    - `git checkout new-feature`
    - `git merge master`
    *merge conflict happens in index.js we fix it*
    - `git add index.js`
    - `git commit -m "fix merge conflict"`
    - `git push origin new-feature`

---

### Javascript
- Using 
	-   `.map()`
	-   `.filter()`
	-   `.find()`

> 	-   `.replace()`

Reviewer: It might be a tiny bit confusing that this is the only non-array method in this list

- [ ] What are error-first callbacks, and why is it important to follow that pattern in your own code?
- Take the following code:
```js

const getUsernameFromDatabase = (email, callback) => {
  const database = db.get("data")
  const user = database.find(user => user.email === email)
  const username = user ? user.username : undefined
  if (!username){
      callback(new Error('No user found'))
  }
  callback(null, username)
}
```
- [ ] Write a code which would call the above function and:
    - [ ] if there is an error, it should `console.log` to the user "Sorry there was a problem"
    - [ ] if it finds a user it should `console.log` to the user "Marhaba {username}!"
- [ ] Finish writing the below function so that when it inserts data into the database it uses the second argument as an *error first* callback. If the insert is successful it should send back `true` in the callback (remembering it is error first).
```js
const addUser = (data, cb) => {
    db_connection.query('INSERT INTO users (name, email,password) VALUES ($1, $2, $3)', [name, email, password], (error) => {
    // write your code here
    })
}
```

> - [ ] Hard challenge! [Parrallel Functions]()

Reviewer: ```suggestion
- [ ] Hard challenge! [Parallel Functions]()
```

Should this link to the challenge?

Author: @m4v15 which challenge did you have in mind for this?

Reviewer: Probably this monster? https://github.com/foundersandcoders/master-reference/blob/master/coursebook/week-3/morning-challenge.md

I didn't realise it wasn't in a repo of its own

Reviewer 2: Ah sorry, yes it would have been this one

- [ ] using supertest, test each of the routes on your server:
	- [ ] check for status codes
	- [ ] check the body response is correct
- [ ] what is the difference between `querystring.parse` and `url.parse`?
- [ ] what is the output from `querstring.stringify({search:'javascript'})` (hint: RUN THIS AND FIND OUT!)
- [ ] what is the output from `url.parse('https://www.codewars.com/users/m4v15/completed')` (hint: RUN THIS AND FIND OUT!)
- If I have the following file structure:
```
-public
  -index.html
  -index.js
  -index.css
-src
  -database
    -db_connect.js
  -app.js
  -server.js
```
- [ ] What file path should I use to `require` the `db_connect` file if I am in `app.js`
- [ ] What is the file path to `index.html` from `app.js`

---

### Databases
You should know:
- [ ] how to connect to a database using `node-postgres (pg)`
- [ ] how to build a database
- [ ] why we use different database url's for production, development and testing
- [ ] If you have a `pg` database connection, `dbConnection` how do you use the `.query` method to query the database? 

> 	- [ ] how should you pass in variables?

Author: To clarify, is this an alternative to both lines 191 and 192? If not, I think 191 needs rephrasing

> 	- [ ] how should you pass in variables?

Reviewer 3: My suggestion was a replacement for L192 🙂. L191 could also do with a reword.

-  **Challenge**:
    - [ ] make a database of users and books, where each user can have multiple books and each book can be owned by multiple users (many to many)
    - [ ] write a small node script that will return the books of a particular user
    - [ ] you will need:
        - a SQL build script
        - a database connection javascript file
        - a database build javascript file
        - a config.env file
        - a app.js javascript file that has a function that takes a name of a user (a string) and queries the database, returning the name of the books that user has
        - Hint:
		    - build script should have 3 tables: users (id, username, location), books (id, name), userbooks(user_id, book_id)

---

### Deploying
- [ ] how to set up travis
	- [ ] How to make travis run your tests
	- [ ] How to make travis run something else (eg a linting script)

> 	- [ ] **Bonus hard check** how to make travis build it's own database for testing on....

Reviewer: ```suggestion
	- [ ] **Bonus hard check** how to make travis build its own database for testing on....
```

- [ ] How to set your app up on heroku
	- [ ] If your app heroku deploy doesn't work, what can you do to find out the problem
	- [ ] How do you provision a database on heroku
	- [ ] How do you add environment variables on heroku
- [ ] What should we store in `config.env` file? Why?
	- [ ] Do the variable values in my `config.env` file have to be the same as their values on heroku?

---

### Auth

- [ ] how to hash and compare using bcrypt
- [ ] how to save and read a cookie on the server
- [ ] how to write a JWT on the server
- [ ] why is a JWT secure?
- [ ] how to secure a route and allow people with a good jwt through

---

### Express

- [ ] Convert the server from the node section to an express server. 
    - [ ] Use express router
    - [ ] add a logger.
    - [ ] You should be putting post requests on the `req.body`
- [ ] What is express middleware? 

---

### Handlebars

> - [ ] Why can server side rendering be useful

Reviewer 3: Missing "?"

- [ ] What is handlebars?
- [ ] Write a small express server that uses a handlebars template to render a user profile page, which show's a user's name, age, location and a list of hobbies, all of which should be passed into the template as `context`. Make endpoints on the server which uses this template for 5 different users (you need to make up the user data!)

---

### Coding Best Practices
- [ ] **eslint**
	- [ ] Why is linting important? Why should all of your code have the same style?
	- [ ] Install eslint on an old project of yours, and use the airbnb style guide. Try to fix all of the problems.

> 	- [ ] Now do the same but use the `standard` style guide - what's the difference?

Reviewer 3: This sounds super boring 😅

How about getting students to create their own linter config.

Suggestion:
- create your own eslint config with all your favorite rules
	- create a `.eslintrc.json` in root
	- copy the following code into your file
```json
{
    "rules": {
        "if-curly-formatting": 1
    }
}
```
	- add the eslint rules you would like to enforce. You can find them all in [eslints docs](https://eslint.org/docs/rules/)
	- stuck? further details on eslint configuration [can be found here](https://eslint.org/docs/user-guide/configuring)

Author: Are you suggesting this as a replacement to lines 251 and 252? Or as an addition?

Reviewer 3: As a replacement for:

> - [ ] Now do the same but use the `standard` style guide - what's the difference?

- [ ] **Prettier** set up prettier on your editor

> 	- [ ] set it to follow the rules in an eslint.config file if one is in the repo

Reviewer: Pretty sure that's not a valid eslint config filename. The docs say it checks for these (in this order):

1. `.eslintrc.js`
1. `.eslintrc.yaml`
1. `.eslintrc.yml`
1. `.eslintrc.json`
1. `.eslintrc`
1. `package.json`

	- [ ] set it to format on save
- [ ] how should you debug your code if things are going wrong? Give some different steps

> - [ ] You **need** to get very comfortable typing on an english keyboard with both hands. Having speed to try things out will help **alot** when you don't understand things.

Reviewer: ```suggestion
- [ ] You **need** to get very comfortable typing on an english keyboard with both hands. Having speed to try things out will help **a lot** when you don't understand things.
```
